### ISSUE TITLE
[incubator/jaeger] Deploy jaeger all-in-one image as default

### ISSUE THREAD

**Issue by objectiser** [helm/charts#5677]
**Is this a BUG REPORT or FEATURE REQUEST?** (choose one): FEATURE REQUEST

**Version of Helm and Kubernetes**: n/a

**Which chart**: incubator/jaeger

**What happened**:
Currently the default deployment of the chart creates a production deployment of Jaeger based on a Cassandra cluster.

**What you expected to happen**:
Jaeger also provides an "all-in-one" image which combines the capabilities of the individual binaries (agent, query+u and collector) into a single binary using an in-memory storage implementation. This image is only intended for demo/dev/test purposes.

We are considering decommissioning the current templates from the https://github.com/jaegertracing/jaeger-kubernetes repo, so users would only use this helm chart to install jaeger into kubernetes - but this means it would also need to offer a way to install the "all-in-one" image, and it seems like this would be the ideal default installation as it is fast.

The issue is that this would change the default behaviour - but it would make the choice of storage now explicit, rather than one (cassandra) being deployed as default.

Ideally the agent should be deployed as a daemonset to be consistent with the production deployments.

cc @dvonthenen @mikelorant @pavelnikolov

**Comment by pavelnikolov** [comment]
I can fix this. As it is a breaking change, it would require major version upgrade. 
Do you need it urgently?

**Comment by objectiser** [comment]
@pavelnikolov No its not urgent - initially just checking if there were any objections to the idea.

Any idea of a timeframe? Just so we can plan for migration from the templates.

**Comment by davidvonthenen** [comment]
@objectiser @pavelnikolov  The addition of the functionality can definitely be a good thing, but question whether or not it should be the default behavior.

I guess it really comes down to what the primary goal of the chart is... 
1) if the chart is primarily to enable demos, then this makes sense.
2) if it's to deploy a test/dev environment, that maybe the all-in-one is suitable. However, if you are going to have a dev/test environment, you might want to actually deploy a production-like environment.
3) if we want to support production type workloads, then I think it shouldn't be the default.

My take on it is to have the option to deploy the all-in-one but have it be an option and not be the default. The chart by default uses scratch space for disk and that disk space is reclaimed once the chart is destroyed. That is pretty close to using in memory without being in memory. We might want to loop in some of the other people who maintain Jaeger to chime in with their 2 cents.

**Comment by pavelnikolov** [comment]
I’ll try to do it next week or the week after.

**Comment by pavelnikolov** [comment]
I have no objection and no strong opinion here. I happy for the all-in-one to be default, since any production deployment would include at least 1 values file to override values (we use 6 values files, some of which change based on environment). 
Also the only time you use:
```
helm install incubator/jaeger
 ```
Is when you are developing/testing. 
In all other cases (ideally) you will have some kind of CI/CD pipeline in place.

**Comment by objectiser** [comment]
@dvonthenen Understand your points - don't have a strong opinion one way or the other. Main reason for suggesting it as default is that in other (production) cases it is likely that user would want to specify their preferred storage and provide additional customisation options. But main requirement is to have the all-in-one available as an option.

**Comment by pavelnikolov** [comment]
If you guys agree on default vs not default, I’ll send a PR soon.

**Comment by objectiser** [comment]
@yurishkuro @jpkrohling @pavolloffay Any preference for whether chart should install all-in-one as the default?

**Comment by pavolloffay** [comment]
I don't know how people are using the chats. If it is for the production or just to try things. So I don't have a strong opinion. More important is if the chart supports all-in-one.

all-ine-one can be a sensible default. Then ES or cassandra should require some configuration.

**Comment by jpkrohling** [comment]
+1 to have the all-in-one as the default. For a production setup, I would expect the operator to read the docs and make a decision about the proper backing storage for their use case, so, they would need to dig into the configuration options anyway.

**Comment by pavelnikolov** [comment]
Ok, I'll try to send a PR some time next week. I will:

- use the `all-in-one` as default
- bump the major version
- update the documentation

**Comment by pavelnikolov** [comment]
@unguiculus any preference?

**Comment by objectiser** [comment]
@pavelnikolov When adding the all-in-one support, just want to make sure you will be using a daemonset for the agent? The current template in jaeger-kubernetes repo doesn't use a daemonset, but was something we were considering changing. So would be good to sort that out as part of including it in the helm chart.

**Comment by pavelnikolov** [comment]
Ok, I'll make sure that the agent is a daemonset and the rest of the components are in the all-in-one deployment.

**Comment by mikejk8s** [comment]
I'm testing these charts out now so I figured I'd provide consumer input. I do like the idea of all-in-one but with the option to deploy to external storage backends for prod use, that greatly simplifies the project. 

Having all 3 options available would be very nice and having the options built right in to fan out a complete cluster or ONE Cass/ES node would be fantastic and has sort of been my biggest hangup.

1. Deploying this w/ Elasticsearch on my laptop via minikube has been painful. I'm not sure what's up right now but I can't get any of the fabric8 discovery to connect the masters. I've used this in staging/prod envs so maybe this is a minikube limitation somewhere. That's another issue though so I won't go into that much more here. Also I've had to dig through the subcharts and manually set minimum node requirements because on minikube I don't need a complete cluster. These vars are not built into the root jaeger chart and everything assumes a minimum of 2 master/data nodes. 
2. Deploying this w/ Cassandra is far less painful and a nice way to get this up and running very quickly. Unfortunately this takes so long to get going I had to set higher deadline seconds so that the Cassandra-schema job doesn't fail and have to be re-initiate manually.
3. In-Memory; In the end while I'm POC'ing this I'd love to not have to run any JVM containers or as few as possible as they take an obscene amount of time to terminate in any k8s environment (laptop, servers). I'm about to test setting the storage to memory but the values.yaml states the two options are cassandra and elasticsearch so we'll see if that works and where we go from there.

So, TLDR; I'd love all 3. I'd like memory for quick POC work and to throw to developers that are locally tracing via minikube, and the other two are great to quickly get things going on a real cluster.

As a user I walked into these charts completely blind other than knowing that CS or ES would be a req in production. Our use-case right now only requires in-memory as we test just getting our services to properly communicate w/ jaeger in general.

Thanks for all your work.

**Comment by pavelnikolov** [comment]
I’m sorry for the delay on this. I’ll start working on the all-in-one very soon.

**Comment by jpkrohling** [comment]
@mikejk8s You have no idea how valuable your feedback is: thank you for sharing it!

**Comment by maeb** [comment]
We use the jaeger chart in production but it would be nice to have an option to use the all-in-one image in dev environment (minikube).
